Add a new particle effect system

asarium opened the pull request:
This allows to define generic particle effects and assign them to specific events. The system uses the `ParticleEffect` class for managing the values of a single effect and the  `ParticleSource` class for managing the properties of a single source of particles.

When an event occurs that should create particles the assigned particle effect will be used to create a particle source. The code that created the source then needs to set the properties of the source. Once that is done the particle manager will process each particle source each frame. In this step the assigned particle effect uses the properties of the particle effect to create new particles according to its own configuration.

This system is similar to a script I wrote a few years ago but unlike that one, this system doesn't kill the performance. This is achieved by strictly separating the properties of effects and sources. Sources only contain a fixed amount of values which means they can be stored in contiguous memory. Because of this, effects are strictly-stateless and can't store values inside a source. A possible future optimization could be to sort the sources vector according to the used effect to improve instruction cache hit rate.

Major changes outside the particle system: `weapon_info` now has fields that turn it into a non trivial struct so `memset` can't be used anymore. To fix that, I introduced `weapon_info::reset()` which repaces the existing `init_weapon_entry` function which initializes all entries to a default value.

MageKing17 commented:
There seem to be quite a few unrelated changes in the commit, making it hard to tell exactly which are related to the new particle effect system.

SamuelCho commented:
Other than the cmake stuff, most of this makes sense and looks good to me. It was too bad we had to cut this from Diaspora so it's nice to see it back and this time tightly integrated in the engine.

asarium commented:
> There seem to be quite a few unrelated changes in the commit, making it hard to tell exactly which are related to the new particle effect system.

I thought I fixed all of those. I reverted the changes to sexp.cpp. The only other mostly unrelated changes I see are in pstypes.h and hudmessage.cpp. I could submit those as separate PRs but those are tiny changes which I didn't want to bother separating from this commit.

MageKing17 commented on code/weapon/weapon.h:
```diff
@@ -418,23 +435,19 @@ typedef struct weapon_info {
-	float piercing_impact_particle_back_velocity;
-	float piercing_impact_particle_variance;
+	// Particle effect for the various states, WeaponState::NORMAL is the state for the whole lifetime, even for missiles
+	SCP_map<WeaponState, particle::ParticleEffectIndex> state_effects;
```
Is it important that this be an ordered map?

asarium replied:
No, I'll change it to an unordered map. That code is from a time when unordered_map wasn't a thing yet...

MageKing17 commented on code/particle/ParticleManager.cpp:
```diff
@@ -0,0 +1,336 @@
+}
+
+void parseConfigFiles() {
+	parse_modular_table("*-part.tbm", parseCallback);
```
This would be (AFAICR) the first modular table with an identification suffix that long, as well as the first without a corresponding `.tbl` file.

The-E replied:
There should definitely be a table (particles.tbl?) in addition to the tbm, even if that tbl is never used by anyone (like scripting.tbl)

asarium replied:
Why should there be a main table even if it isn't used by anyone? I guess it helps the documentation but at least for the scripting.tbl case it has caused some problems where different mods have conflicting tables.

Adding a table is only a one line change but I don't see a good reason why a table should exist.

asarium commented:
Now I remember why I used a `SCP_map`. WeaponState is an enum class which means that it is not hashable. I'll have to do some template magic to fix that.

asarium commented:
CI builds have been fixed. The only remaining point is the main particle effects table.
